iwyu.cc

```diff
@@ -801,6 +812,18 @@ class BaseAstVisitor : public RecursiveASTVisitor<Derived> {
     const Type* parent_type = expr->getAllocatedType().getTypePtrOrNull();
     // 'new' calls operator new in addition to the ctor of the new-ed type.
     if (FunctionDecl* operator_new = expr->getOperatorNew()) {
+      // If the placement-new call makes use of default arguments, select
```

Comment:
I find these for loops really hard to read. I suspect a less compact version would be more straighforward:
```
for (unsigned i = 0; i < expr->getNumPlacementArgs(); ++i) {
  if (isa<CXXDefaultArgExpr>(expr->getPlacementArg(i)) {
    CHECK_((i + 1) < expr->getNumPlacementArgs());
    operator_new = GetRedeclSpecifyingDefArg(i + 1, operator_new);
    break;
  }
}
```

That would also allow you to get rid of the "allow-out-of-bounds" thing in `GetRedeclSpecifying...`

Reply:
If you declare an allocation function like this:
```cpp
void* operator new(std::size_t size, int a, int b);
```
then for a call like `new(1, 2) int` the first parameter (`size`) is initialized with the result of `sizeof(int)`, the second parameter (`a`) is initialized with the first argument expression (`1`), and the third parameter (`b`) is initialized with the second argument (`2`): https://godbolt.org/z/aeGc6eME1

I left a comment below, but maybe it is not clear enough. Also, there is a couple of test cases.